lua/network_promises/async.lua

```diff
@@ -0,0 +1,98 @@
+local promiseReturn
+
+promiseReturn = function( f, coInput, state, ownPInput, debugInfo )
+    return function( ... )
+        local ownP = ownPInput
+        local isRoot = ownP == nil
+        local co = coInput
+        if isRoot then
+            ownP = promise.new()
+            co = coroutine.create( f )
+        end
+
+        local ret
+        if state ~= nil then
+            ret = { coroutine.resume( co, state, ... ) }
+        else
+            ret = { coroutine.resume( co, ... ) }
+        end
+        local ok = ret[1]
+        if ok then
+            if ret[2] and type( ret[2] ) == "table" and ret[2].next then
+                ret[2]:next( promiseReturn( f, co, true, ownP, debugInfo ), promiseReturn( f, co, false, ownP, debugInfo ) )
+            else
+                table.remove( ret, 1 )
+                -- Delay to escape any enclosing pcalls
+                timer.Simple( 0, function()
+                    if state ~= false then
+                        ownP:resolve( unpack( ret ) )
+                    else
+                        ownP:reject( unpack( ret ) )
+                    end
+                end )
+            end
+        else
+            -- Delay to escape any enclosing pcalls
+            timer.Simple( 0, function()
+                ownP:reject( ret[2], "\nContaining async function defined in " .. debugInfo.short_src .. " at line " .. debugInfo.linedefined .. "\n" )
+            end )
+        end
+
+        if isRoot then
+            return ownP
+        end
+    end
+end
```

Comment:
While understandably so, this chunk of code would be very difficult to work with for someone who was coming into this project fresh.

Some better spacing, more well named variables, and extracting some nested chunks out to different functions could go a long way here, I think.

Reply:
Oh boy, this'll be a fun refactor